src/rust_api.rs

```diff
@@ -0,0 +1,551 @@
+    V,
+}
+
+impl TryFrom<usize> for PlanarImageComponent {
+    type Error = Rav1dError;
+    fn try_from(index: usize) -> Result<Self, Self::Error> {
+        match index {
+            0 => Ok(PlanarImageComponent::Y),
+            1 => Ok(PlanarImageComponent::U),
+            2 => Ok(PlanarImageComponent::V),
+            _ => Err(Rav1dError::InvalidArgument),
+        }
+    }
+}
+
+impl From<PlanarImageComponent> for usize {
+    fn from(component: PlanarImageComponent) -> Self {
+        match component {
+            PlanarImageComponent::Y => 0,
+            PlanarImageComponent::U => 1,
+            PlanarImageComponent::V => 2,
+        }
+    }
+}
+
+/// A single plane of a decoded frame.
+///
+/// This can be used like a `&[u8]`.
+#[derive(Clone)]
+pub struct Plane(Picture, PlanarImageComponent);
+
+impl AsRef<[u8]> for Plane {
+    fn as_ref(&self) -> &[u8] {
+        let (stride, height) = self.0.plane_data_geometry(self.1);
+        // SAFETY: both stride and height can't be negative, the `stride` and `height` methods panic if they are so there's no undefined behaviour
+        unsafe {
+            slice::from_raw_parts(
+                self.0.plane_data_ptr(self.1) as *const u8,
+                (stride * height) as usize,
+            )
+        }
+    }
+}
+
+impl Deref for Plane {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+/// Number of bits per component.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct BitsPerComponent(pub u8);
+
+impl TryFrom<u8> for BitsPerComponent {
+    type Error = Rav1dError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(BitsPerComponent(8)),
+            1 => Ok(BitsPerComponent(10)),
+            2 => Ok(BitsPerComponent(12)),
+            _ => Err(Rav1dError::InvalidArgument),
+        }
+    }
+}
+
+impl Picture {
+    /// Stride in pixels of the `component` for the decoded frame.
+    pub fn stride(&self, component: PlanarImageComponent) -> u32 {
+        let s = match component {
+            PlanarImageComponent::Y => 0,
+            _ => 1,
+        };
+        self.inner.stride[s].try_into().unwrap()
+    }
+
+    /// Raw pointer to the data of the `component` for the decoded frame.
+    pub fn plane_data_ptr(&self, component: PlanarImageComponent) -> *mut c_void {
+        let index: usize = component.into();
+        self.inner.data.as_ref().unwrap().data[index]
+            .as_byte_mut_ptr()
+            .cast()
+    }
+
+    /// Plane geometry of the `component` for the decoded frame.
+    ///
+    /// This returns the stride and height.
+    pub fn plane_data_geometry(&self, component: PlanarImageComponent) -> (u32, u32) {
+        let height = match component {
+            PlanarImageComponent::Y => self.height(),
+            _ => match self.pixel_layout() {
+                PixelLayout::I420 => (self.height() + 1) / 2,
```

Comment:
```suggestion
                PixelLayout::I420 => self.height().div_ceil(2),
```

Handles the case where `height = u32::MAX` correctly, too.